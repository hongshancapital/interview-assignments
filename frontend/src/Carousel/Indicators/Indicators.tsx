import { FC } from 'react';
import styles from './Indicators.module.scss';

type IndicatorsProps = {
  currentActiveIndex: number;
  length: number;
  duration: number;
};

export const Indicators: FC<IndicatorsProps> = ({
  currentActiveIndex,
  length,
  duration,
}) => {
  const indicators = new Array(length).fill(null);
  return (
    <div className={styles.indicators}>
      {indicators.map((_, index) => {
        const isActive = currentActiveIndex === index;
        return (
          <div
            className={styles.indicator}
            aria-label="indicator"
            role="progressbar"
            // There is no insert/delete/resort case for indicators
            // so use index as key is fine.
            key={index}
          >
            <span
              className={styles.indicatorProgress}
              aria-label="indicator progress"
              style={{
                opacity: isActive ? 1 : 0,
                animation: isActive
                  ? `${styles.progress} ${duration}ms linear`
                  : 'none',
              }}
            ></span>
          </div>
        );
      })}
    </div>
  );
};
