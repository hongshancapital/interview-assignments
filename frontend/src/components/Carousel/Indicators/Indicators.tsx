import { FC } from 'react';
import styles from './Indicators.module.scss';

type IndicatorsProps = {
  activeIndex: number;
  length: number;
  duration: number;
};

/**
 * @Component Indicators
 * @description Carousel indicators, which are progressbar at the bottom of the carousel.
 */
export const Indicators: FC<IndicatorsProps> = ({
  activeIndex,
  length,
  duration,
}) => {
  const indicators = new Array(length).fill(null);
  return (
    <ul className={styles.indicators}>
      {indicators.map((_, index) => {
        const isActive = activeIndex === index;
        return (
          <li
            className={styles.indicator}
            aria-label="indicator"
            role="progressbar"
            // There is no insert/delete/resort case for indicators
            // so use index as key is fine.
            key={index}
          >
            <span
              className={styles.indicatorProgress}
              aria-label="indicator progress"
              style={{
                opacity: isActive ? 1 : 0,
                animation: isActive
                  ? `${styles.progress} ${duration}ms linear`
                  : 'none',
              }}
            />
          </li>
        );
      })}
    </ul>
  );
};
