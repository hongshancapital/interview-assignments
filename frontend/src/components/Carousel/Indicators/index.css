@keyframes indicatorScale {
  0% {
    transform: scaleX(0);
  }
  100% {
    transform: scaleX(1);
  }
}
.carousel-indicators {
  width: 100%;
  height: 4px;
  position: absolute;
  bottom: 6vh;
  display: flex;
  justify-content: center;
}
.carousel-indicator {
  width: 56px;
  height: 4px;
  background: #aaa;
  border-radius: 4px;
  overflow: hidden;
  position: relative;
  cursor: pointer;
}
.carousel-indicator + .carousel-indicator {
  margin-left: 10px;
}
.indicator-content {
  position: absolute;
  height: 100%;
  width: 100%;
}
.indicator-progress {
  position: absolute;
  left: 0;
  top: 0;
  height: 100%;
  background: #fff;
  width: 100%;
  transform: scaleX(0);
  transform-origin: left center;
}
.carousel-indicator-active .indicator-progress {
  animation-name: indicatorScale;
  animation-timing-function: linear;
  animation-fill-mode: forwards;
}
.carousel-indicator-active .indicator-content {
  background: #fff;
}
