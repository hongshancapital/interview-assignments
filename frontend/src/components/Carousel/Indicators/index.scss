@keyframes indicatorScale {
  0% {
    transform: scaleX(0);
  }
  100% {
    transform: scaleX(1);
  }
}
.carousel__indicators {
  width: 100%;
  height: 4px;
  position: absolute;
  bottom: 2vh;
  display: flex;
  justify-content: center;
  .carousel__indicator {
    width: 56px;
    height: 4px;
    background: #aaa;
    border-radius: 4px;
    overflow: hidden;
    position: relative;
    cursor: pointer;
  }
  .carousel__indicator + .carousel__indicator {
    margin-left: 10px;
  }
  .indicator__content {
    position: absolute;
    height: 100%;
    width: 100%;
  }
  .indicator__progress {
    position: absolute;
    left: 0;
    top: 0;
    height: 100%;
    background: #fff;
    width: 100%;
    transform: scaleX(0);
    transform-origin: left center;
  }
  .carousel__indicator-active {
    .indicator__progress {
      animation-name: indicatorScale;
      animation-timing-function: linear;
      animation-fill-mode: forwards;
    }
    .indicator__content {
      background: #fff;
    }
  }
}