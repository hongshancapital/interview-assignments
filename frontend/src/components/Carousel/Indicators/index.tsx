import React, { useState } from "react";
import "./index.css";

type CarouselIndicatorsProps = {
  autoPlay?: boolean;
  activeIndex: number;
  count: number;
  interval: number;
  goto: (index: number) => void;
};

const Indicators = (props: CarouselIndicatorsProps) => {
  const { count, activeIndex, autoPlay, interval, goto } = props;
  const [animationPaused, setAnimationPaused] = useState(false);

  const handleOnclick = (index: number) => goto(index);
  const handleMouseEnter = (index: number) => index === activeIndex && setAnimationPaused(true);
  const handleMouseLeave = (index: number) => index === activeIndex && setAnimationPaused(false);
  const handleAnimationEnd = () => goto((activeIndex + 1) % count);

  return (
    <div className="carousel-indicators">
      {new Array(count).fill(true).map((_, index) => {
        return (
          <div
            key={index}
            onClick={() => handleOnclick(index)}
            onMouseEnter={() => handleMouseEnter(index)}
            onMouseLeave={() => handleMouseLeave(index)}
            className={
              activeIndex === index
                ? "carousel-indicator carousel-indicator-active"
                : "carousel-indicator"
            }
          >
            {autoPlay ? (
              <div
                className="indicator-progress"
                style={{ 
                  animationDuration: `${interval}ms`,
                  animationPlayState: animationPaused ? "paused" : "running",
               }}
                onAnimationEnd={handleAnimationEnd}
              ></div>
            ) : (
              <div className="indicator-content"></div>
            )}
          </div>
        );
      })}
    </div>
  );
};

export default Indicators;
