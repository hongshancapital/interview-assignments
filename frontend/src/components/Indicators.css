.indicators {
    position: absolute;
    bottom: 60px;
    left: 0;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: flex-start;
}

.indicator {
    background-color: #9f9fa1;
    height: 2px;
    width: 60px;
    margin-right: 8px;
}

.indicator:last-child {
    margin-right: 0;
}

.indicator::-webkit-progress-value {
    background-color: white;
}