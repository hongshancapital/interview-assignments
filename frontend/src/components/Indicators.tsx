import React, { useState, useEffect } from "react";
import useRaf from "../hooks/useRaf";
import "./Indicators.css";
import { TSlide } from "./carousel-types";

interface TIndicatorProps {
  onPlayEnded: () => void;
  playSpeed: number;
}

interface TIndicatorsProps extends TIndicatorProps {
  slides: Array<TSlide>;
  currentSlide: number
}

function NormalIndicator() {
  return <progress max="100" value="0" className="indicator" />;
}

function ActiveIndicator(props: TIndicatorProps) {
  const progress = useRaf(props.playSpeed, props.onPlayEnded);
  return <progress max="100" value={`${progress}`} className="indicator" />;
}

function Indicators(props: TIndicatorsProps) {
  const { slides, currentSlide, onPlayEnded, playSpeed } = props;

  return (
    <div className="indicators">
      {slides.map((item, index) => {
        if (index === currentSlide) {
          return (
            <ActiveIndicator
              playSpeed={playSpeed}
              onPlayEnded={onPlayEnded}
              key={`indicator_${index}`}
            />
          );
        }
        return <NormalIndicator key={`indicator_${index}`} />;
      })}
    </div>
  );
}

export default Indicators;
