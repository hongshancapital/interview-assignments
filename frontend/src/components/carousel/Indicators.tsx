import React from "react";
import './style/index.scss'

export interface CarouselIndicatorsProps {
  activeIndex: number,
  count: number,
  animationDuration: number,
}

const Indicators = (props: CarouselIndicatorsProps) => { 

  const { activeIndex, count, animationDuration } = props;
  if (count <= 1) {
    return null;
  }
  return (
    <div className="carousel-indicators__wrap">
      {(new Array(count)).fill('').map((_, index) => (
        <div key={index} className="carousel-indicator">
          <div
            className={index === activeIndex ? 'carousel-indicator__progress' : ''}
            style={{ animationDuration: `${animationDuration}ms` }}
          />
        </div>
      ))}
    </div>
  )
}

export default Indicators;