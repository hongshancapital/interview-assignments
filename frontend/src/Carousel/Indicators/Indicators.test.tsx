import { render, screen } from '@testing-library/react';
import { Indicators } from './Indicators';

describe('Indicators', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <Indicators length={3} currentActiveIndex={0} duration={1000} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('shows the progress and has animation for active indicator', () => {
    render(<Indicators length={3} currentActiveIndex={1} duration={1000} />);

    expect(screen.getAllByLabelText('indicator progress')[1]).toHaveStyle({
      'opacity': 1,
      animation: 'progress 1000ms linear',
    });
  });
  it('hides the progress and set animation to none for inactive indicator', () => {
    render(<Indicators length={3} currentActiveIndex={1} duration={1000} />);

    expect(screen.getAllByLabelText('indicator progress')[0]).toHaveStyle({
      'opacity': 0,
      animation: 'none',
    });
  });
});
