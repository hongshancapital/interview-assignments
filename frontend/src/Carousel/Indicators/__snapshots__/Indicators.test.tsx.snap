// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicators renders correctly 1`] = `
<DocumentFragment>
  <div
    class="indicators"
  >
    <div
      aria-label="indicator"
      class="indicator"
      role="progressbar"
    >
      <span
        aria-label="indicator progress"
        class="indicatorProgress"
        style="opacity: 1; animation: progress 1000ms linear;"
      />
    </div>
    <div
      aria-label="indicator"
      class="indicator"
      role="progressbar"
    >
      <span
        aria-label="indicator progress"
        class="indicatorProgress"
        style="opacity: 0; animation: none;"
      />
    </div>
    <div
      aria-label="indicator"
      class="indicator"
      role="progressbar"
    >
      <span
        aria-label="indicator progress"
        class="indicatorProgress"
        style="opacity: 0; animation: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;
